warc_metadata_sidecar.py

```diff
@@ -0,0 +1,279 @@
+        os.mkdir(archive_dir)
+
+    logging.basicConfig(
+        filename=os.path.join(archive_dir, 'sidecar.log'),
+        level=logging.INFO,
+        format='%(asctime)s %(levelname)s %(message)s',
+    )
+    logging.getLogger(__name__)
+    logging.info('Logging WARC metadata record information for %s', warc_file)
+
+    # Create sidecar filename, adding 'meta' as extension.
+    new_file = os.path.basename(warc_file)
+    meta_file = re.sub(r'warc(\.gz)?$', 'warc.meta.gz', new_file)
+    logging.info('Creating sidecar %s', meta_file)
+    warc_file_path = os.path.join(archive_dir, meta_file)
+
+    # Open the sidecar file to write in the metadata, open the warc file to get each record.
+    with open(warc_file_path, 'ab') as output, open(warc_file, 'rb') as stream:
+        record_count = 0
+        total_records = 0
+        text_mime = 0
+        non_text = 0
+
+        writer = WARCWriter(output, gzip=False)  # TODO: gzip will equal True
+        warc_info = create_warcinfo_payload(new_file, operator, publisher)
+        # Create warcinfo record and write it into sidecar.
+        warcinfo_record = writer.create_warcinfo_record(meta_file, warc_info)
+        writer.write_record(warcinfo_record)
+
+        for record in ArchiveIterator(stream):
+            fido = ExtendFido()
+            total_records += 1
+            url = record.rec_headers.get_header('WARC-Target-URI')
+            payload = io.BytesIO(record.content_stream().read())
+            if record.rec_type not in ['response', 'resource']:
+                continue
+            # The payload is how we find the important info. Skip record if empty.
+            if not payload.read(1):
+                continue
+            if 'text/dns' in record.rec_headers.get_header('Content-Type'):
+                continue
+            record_date = record.rec_headers.get_header('WARC-Date')
+            warcinfo_id = record.rec_headers.get_header('WARC-Warcinfo-ID')
+            warcrecord_id = record.rec_headers.get_header('WARC-Record-ID')
+            # Define specific warc_headers to include in sidecar.
+            warc_dict = {'WARC-Date': record_date, 'WARC-Concurrent-ID': warcrecord_id}
+            if warcinfo_id:
+                warc_dict['WARC-Warcinfo-ID'] = warcinfo_id
+
+            print(url)
+            payload.seek(0)
+            find_mime_and_puid(fido, payload, url)
+
+            puid = '\n{0} {1}'.format(PUID_TITLE, fido.puid) if fido.puid is not None else None
+            if fido.mime:
+                mime_str = '{0} {1}\n{2} {3}\n{4} {5}'.format(
+                            MIME_TITLE, fido.mime,
+                            METHOD_TITLE, fido.matchtype,
+                            RETRIEVED_TITLE, fido.format_tool)
+                string_payload = mime_str
+                if puid:
+                    string_payload += puid
+                text_format_mimes = re.compile(r'(text|html|xml)')
+                if text_format_mimes.search(fido.mime):
```

Comment:
Noting we may end up wanting to skip some text-based mimes like 'text/css'. Are there others? We should discuss with @vphill .